0300-Experimental-Xen-hacking-with-GRUB-parameters.patch

```diff
@@ -0,0 +1,366 @@
++              if (!append_char_to_word (false))
++                {
++                  goto cleanup;
++                }
++              continue;
++            }
++
++          parse_flags ^= PARSER_IN_DOUBLE_QUOTES;
++          continue;
++        }
++
++      if (current_char == ' ')
++        {
++          if (parse_flags & PARSER_IN_SINGLE_QUOTES
++              || parse_flags & PARSER_IN_DOUBLE_QUOTES)
++            {
++              if (!append_char_to_word (false))
++                {
++                  goto cleanup;
++                }
++              continue;
++            }
++
++          append_word_to_list ();
++          continue;
++        }
++
++      if (!append_char_to_word (false))
++        {
++          goto cleanup;
++        }
++    }
++
++  if (current_word_pos > 0)
++    {
++      append_word_to_list ();
++    }
++
++  param_keys = grub_realloc (param_keys, word_list_len * sizeof (char *));
++  param_vals = grub_realloc (param_vals, word_list_len * sizeof (char *));
++  for (grub_size_t i = 0; i < word_list_len; i++)
++    {
++      current_word = word_list[i];
++      current_word_len = grub_strlen (current_word) + 1;
++      char *current_word_eq_ptr = grub_strchr (current_word, '=');
++      if (current_word_eq_ptr)
++        {
++          grub_size_t eq_idx
++              = (grub_size_t)(current_word_eq_ptr - current_word);
++          grub_size_t pre_eq_len
++              = current_word_len - (current_word_len - eq_idx);
++          grub_size_t post_eq_len = current_word_len - eq_idx - 1;
++          if (check_key_is_safe (current_word, pre_eq_len))
++            {
++              param_keys[param_dict_len] = grub_malloc (pre_eq_len + 1);
++              param_vals[param_dict_len] = grub_malloc (post_eq_len + 1);
++              grub_strncpy (param_keys[param_dict_len],
++			    current_word, pre_eq_len);
++              grub_strncpy (param_vals[param_dict_len],
++			    current_word + pre_eq_len + 1, post_eq_len);
++              param_keys[param_dict_len][pre_eq_len] = '\0';
++              param_vals[param_dict_len][post_eq_len] = '\0';
++	      param_dict_len++;
++            }
++        }
++      else
++        {
++          if (check_key_is_safe (current_word, current_word_len - 1))
++            {
++              param_keys[param_dict_len] = grub_malloc (current_word_len + 1);
++              param_vals[param_dict_len] = grub_malloc (2);
++              grub_strncpy (param_keys[param_dict_len],
++			    current_word, current_word_len);
++              param_keys[param_dict_len][current_word_len] = '\0';
++              grub_strcpy (param_vals[param_dict_len], "1\0");
++	      param_dict_len++;
++            }
++        }
++    }
++
++  for (grub_size_t i = 0; i < param_dict_len; i++)
++    {
++      /* 
++       * Find keys that start with "xen_grub_env_" and export them
++       * as environment variables.
++       */
++      if (grub_strlen (param_keys[i]) < (sizeof ("xen_grub_env_") - 1))
```

Comment:
You don't really need this one if both arguments for `grub_strncmp` below are NUL-terminated already (if it's shorter, then terminating NUL earlier in the key won't match `xen_grub_env_` already). And if they wouldn't be NUL-terminated, then `strlen` wouldn't help either, but fortunately they are - you ensure that earlier.

Reply:
kk, that's a good optimization to add.

Reply:
Optimized.